alchemiscale/storage/statestore.py

```diff
@@ -242,138 +242,139 @@ def reset(self):
+            ] = node
+
+        def get_previous_node(node_gufe_key):
+            if (
+                rel_node := previous_nodes.get(
+                    (node_gufe_key, scope.org, scope.campaign, scope.project)
                 )
-                if node_ is None:
-                    subgraph_, node_, scoped_key_ = self._gufe_to_subgraph(
-                        value.to_shallow_dict(),
-                        labels=["GufeTokenizable", value.__class__.__name__],
-                        gufe_key=value.key,
-                        scope=scope,
-                    )
-                    self.gufe_nodes[
-                        (value.key, scope.org, scope.campaign, scope.project)
-                    ] = node_
-                subgraph = (
-                    subgraph
-                    | Relationship.type("DEPENDS_ON")(
-                        node,
-                        node_,
-                        attribute=key,
-                        _org=scope.org,
-                        _campaign=scope.campaign,
-                        _project=scope.project,
+            ) is None:
+                raise ValueError("Possibly corrupt keyedchain")
+            return rel_node
+
+        def update_relationships(node_a, node_b, **kwargs):
+            rel = Relationship.type("DEPENDS_ON")(
+                node_a,
+                node_b,
+                _org=scope.org,
+                _campaign=scope.campaign,
+                _project=scope.project,
+                **kwargs,
+            )
+            relationships.append(rel)
+
+        def handle_list(node, key, values):
+            if isinstance(values[0], (int, float, str)) and all(
+                (isinstance(x, type(values[0])) for x in values)
+            ):
+                node[key] = values
+                return
+            # list of gufe key: [{":gufe-key:": ...}, {":gufe-key:": ...}, {":gufe-key:": ...}]
+            elif all(map(is_gufe_dict, values)):
+                for i, x in enumerate(values):
+                    rel_key = x[":gufe-key:"]
+                    update_relationships(
+                        node, get_previous_node(rel_key), attribute=key, index=i
                     )
-                    | subgraph_
-                )
+            else:
+                node[key] = json.dumps(values, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
 
+        def handle_dict(node, key, dct):
+            # e.g. {'ligand': {':gufe-key:': 'SmallMoleculeComponent-abc123'}}
+            if all(map(is_gufe_dict, dct.values())):
+                for k, v in dct.items():
+                    rel_key = v[":gufe-key:"]
+                    update_relationships(
+                        node, get_previous_node(rel_key), attribute=key, key=k
+                    )
             else:
-                node[key] = value
+                node[key] = json.dumps(dct, cls=JSON_HANDLER.encoder)
+                node["_json_props"].append(key)
```

Comment:
Nitpick: can we put these methods in the same order as they appear in `process_keyed_dict`?

Reply:
sure